### PR TITLE
[Data] Add support for objects to Arrow blocks

### LINKED ISSUE #45235: [Data] Can't return array-like data from UDF if batch contains unsupported type
### What happened + What you expected to happen

I returned array data from my UDF, but I got an error saying that arrays must be 1-dimensional:
```
(ReadRange->MapBatches(f) pid=69903) Could not construct Arrow block from numpy array; encountered values of unsupported numpy type `17` in column named 'unsupported', which cannot be casted to an Arrow data type. Falling back to using pandas block type, which is slower and consumes more memory. For maximum performance, consider applying the following suggestions before ingesting into Ray Data in order to use native Arrow block types:
(ReadRange->MapBatches(f) pid=69903) - Expand out each key-value pair in the dict column into its own column                  
(ReadRange->MapBatches(f) pid=69903) - Replace `None` values with an Arrow supported data type                                
(ReadRange->MapBatches(f) pid=69903)                                                                                          
Running 0:   0%|                                                                                       | 0/20 [00:00<?, ?it/s]2024-05-09 17:23:05,898(ERROR streaming_executor_state.py:455 -- An exception was raised from a task of operator "ReadRange->MapBatches(f)". Dataset execution will now abort. To ignore this exception and continue, set DataContext.max_errored_blocks.
                                                                                                                             2024-05-09 17:23:05,916  ERROR exceptions.py:73 -- Exception occurred in Ray Data or Ray Core internal code. If you continue to see this error, please open an issue on the Ray project GitHub page with the full stack trace below: https://github.com/ray-project/ray/issues/new/choose
ray.data.exceptions.SystemException

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/balaji/Documents/GitHub/ray/1.py", line 30, in <module>
    ray.data.range(100).map_batches(f).materialize()
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/dataset.py", line 4541, in materialize
    copy._plan.execute(force_read=True)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/exceptions.py", line 86, in handle_trace
    raise e.with_traceback(None) from SystemException()
ray.exceptions.RayTaskError(ValueError): ray::ReadRange->MapBatches(f)() (pid=69903, ip=127.0.0.1)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/arrow_block.py", line 210, in numpy_to_block
    col = ArrowTensorArray.from_numpy(col, col_name)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/balaji/Documents/GitHub/ray/python/ray/air/util/tensor_extensions/arrow.py", line 376, in from_numpy
    raise e
  File "/Users/balaji/Documents/GitHub/ray/python/ray/air/util/tensor_extensions/arrow.py", line 336, in from_numpy
    pa_dtype = pa.from_numpy_dtype(arr.dtype)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pyarrow/types.pxi", line 5164, in pyarrow.lib.from_numpy_dtype
  File "pyarrow/error.pxi", line 154, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 91, in pyarrow.lib.check_status
pyarrow.lib.ArrowNotImplementedError: Unsupported numpy type 17

During handling of the above exception, another exception occurred:

ray::ReadRange->MapBatches(f)() (pid=69903, ip=127.0.0.1)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/execution/operators/map_operator.py", line 410, in _map_task
    for b_out in map_transformer.apply_transform(iter(blocks), ctx):
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/execution/operators/map_transformer.py", line 393, in __call__
    add_fn(data)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/output_buffer.py", line 48, in add_batch
    self._buffer.add_batch(batch)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/delegating_block_builder.py", line 38, in add_batch
    block = BlockAccessor.batch_to_block(batch)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/block.py", line 380, in batch_to_block
    return pd.DataFrame(dict(batch))
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/balaji/anaconda3/envs/ray/lib/python3.11/site-packages/pandas/core/frame.py", line 664, in __init__
    mgr = dict_to_mgr(data, index, columns, dtype=dtype, copy=copy, typ=manager)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/balaji/anaconda3/envs/ray/lib/python3.11/site-packages/pandas/core/internals/construction.py", line 493, in dict_to_mgr
    return arrays_to_mgr(arrays, columns, index, dtype=dtype, typ=typ, consolidate=copy)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/balaji/anaconda3/envs/ray/lib/python3.11/site-packages/pandas/core/internals/construction.py", line 118, in arrays_to_mgr
    index = _extract_index(arrays)
            ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/balaji/anaconda3/envs/ray/lib/python3.11/site-packages/pandas/core/internals/construction.py", line 653, in _extract_index
    raise ValueError("Per-column arrays must each be 1-dimensional")
ValueError: Per-column arrays must each be 1-dimensional
```

### Versions / Dependencies

6a266db05a8228a7d66cb88e8b1ced862376edf8

### Reproduction script

```python
import numpy as np

import ray


class UnsupportedType:
    pass


def f(batch):
    batch_size = len(batch["id"])
    return {
        "array": np.zeros((batch_size, 32, 32, 3)),
        "unsupported": [UnsupportedType()] * batch_size,
    }


ray.data.range(100).map_batches(f).materialize()
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by terraflops1048576**
## Why are these changes needed?

Currently, Ray does not support blocks/batches with objects and multi-dimensional arrays in different columns. This causes Ray Data to throw exceptions when these are provided because:
1. Since there's an arbitrary object in the batch, the Arrow block format fails with ArrowNotImplemented with dtype 17. This falls back to `return pd.DataFrame(dict(batch))` in `BlockAccessor.batch_to_block`.
2. However, this particular DataFrame constructor does not support columns with numpy.ndarray objects, so it throws the exception listed in the linked issue.

This change enables Python object storage in the Arrow blocks by defining an Arrow extension type that simply represents the Python objects as a variable-sized large binary. I suppose the alleged performance benefits listed in the comments are an extra benefit.

I'm not sure that this is the correct approach or that I've properly patched all of the places, so some help would be appreciated!

## Related issue number

Resolves #45235
## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/tests/test_arrow_serialization.py
Comment (terraflops1048576) [inline python/ray/data/tests/test_arrow_serialization.py:545]: I just needed an object that wasn't subject to the vagaries of the importing from the test module, so I chose this.

File: python/ray/air/util/object_extensions/arrow.py
Comment (bveeramani) [inline python/ray/air/util/object_extensions/arrow.py:13]: What happens if we use a lower version of PyArrow? Which APIs do we need?

Comment (terraflops1048576) [reply python/ray/air/util/object_extensions/arrow.py:13]: Subclassing extension scalars is only available in PyArrow 9.0. This is needed for the code:
```
@PublicAPI(stability="alpha")
class ArrowPythonObjectScalar(pa.ExtensionScalar):
    def as_py(self) -> typing.Any:
        if not isinstance(self.value, pa.LargeBinaryScalar):
            raise RuntimeError(
                f"{type(self.value)} is not the expected LargeBinaryScalar"
            )
        return pickle.load(pa.BufferReader(self.value.as_buffer()))
```
that allows all of the as_py conversions in the array.

Comment (anyscalesam) [reply python/ray/air/util/object_extensions/arrow.py:13]: @bveeramani WDYT; OK (if so please resolve this conflict).

**Comment by anyscalesam** [comment]
@raulchen is this ready to merge?

**Comment by raulchen** [comment]
sorry, I got disruptions last week. doing the 2nd pass of review now. 
Also, could you resolve the conflicts when you have a chance? thanks

File: python/ray/data/_internal/arrow_ops/transform_pyarrow.py
Comment (raulchen) [inline python/ray/data/_internal/arrow_ops/transform_pyarrow.py:78]: can you explain what this change is doing?

**Review by raulchen — COMMENTED** [review]
Looks good to me at the high level.
Apologize for my late review. Also none of the active Ray Data maintainers are familiar with the pyarrow extension. it'd be appreciated if you can help explain the questions and add more comments. 
thanks.

File: python/ray/air/util/object_extensions/arrow.py
Comment (terraflops1048576) [reply python/ray/air/util/object_extensions/arrow.py:19]: Actually, I'm not sure. I just borrowed these version checks from the existing ArrowTensorArray implementation, which also passes if the PYARROW_VERSION is None. It turns out the way I've done it is buggy, so I'll revise it. I can also defensively just make this check fail if PYARROW_VERSION is None.

The more general question is whether we would be open to bumping the minimum PyArrow version to 9, which is almost 2 years old at this point. Currently, it is at 6.0.1, which appears to be almost 3 years old.

File: python/ray/air/util/object_extensions/arrow.py
Comment (terraflops1048576) [reply python/ray/air/util/object_extensions/arrow.py:25]: Will do

File: python/ray/data/_internal/arrow_block.py
Comment (terraflops1048576) [reply python/ray/data/_internal/arrow_block.py:153]: I believe this is the way that the fallback to Pandas is calculated; in any case, this code is outdated after the merge conflict because they have now all been coalesced into ArrowConversionError.

It is possible to check that all of the objects inside are of some Arrow type, but I believe this is quite cumbersome -- you have to manually check whether it is convertible to any PyArrow type, and you have to check all of the elements. It's easier to try conversion and see if it works.

In any case, the conversion into the ArrowPythonObjectArray should never fail if everything is pickleable, and I assume this is better than falling back to Pandas.

File: python/ray/data/_internal/arrow_block.py
Comment (terraflops1048576) [reply python/ray/data/_internal/arrow_block.py:245]: The current code looks at any column which has a numpy dtype of object and checks if it can be converted into a list of tensors. If it can't be, then we fall back to ArrowPythonObjectArray. This happens when the numpy array has a dtype of object because you have custom objects in there.

File: python/ray/data/tests/test_arrow_block.py
Comment (terraflops1048576) [reply python/ray/data/tests/test_arrow_block.py:62]: Ah yes, this used to be a test that checked the fallback behavior to Pandas blocks, but these changes are supposed to remove that fallback behavior for the case given. I cut the test from where it originally was and pasted it here, but I must've forgotten to delete the comment.

File: python/ray/data/_internal/arrow_ops/transform_pyarrow.py
Comment (terraflops1048576) [reply python/ray/data/_internal/arrow_ops/transform_pyarrow.py:78]: Basically this is a refactor of the unify_schemas function, which serves a similar purpose to the PyArrow unify_schemas. Its job is to take the schemas of the tables to be concatenated and produce a "unified" schema which contains the correct types for all of the columns of the constituent tables.

The previous code checked only the first table in the tables to concatenate, so it would behave weirdly if the first table weren't a tensor type and there was one further along by creating a schema where the tensor types were just erased (the error would be caught elsewhere, though).

However, for the object array, we need to support the case where the first table has a column of type, say, int64, and the second table has a column of type SomeObject. The old code would simply set the type to int64, which would cause an error. The refactored code will ensure that the int64s get pickled (yes, this is not very efficient, but this is the only way to support  So this new code scans all of the table schemas, finds which columns have any table that contain tensors or objects, and handles them accordingly.

This is basically type promotion for the ArrowTensorArray and the ArrowVariableShapedTensorArray and likewise for the ArrowPythonObjectArray and other types. Should we set the promotion options differently too?

File: python/ray/data/tests/test_transform_pyarrow.py
Comment (terraflops1048576) [inline python/ray/data/tests/test_transform_pyarrow.py:188]: I'm not sure if these actually get run in CI if I put the skipif and require PyArrow version >= 9.0.0. Can someone check?

**Comment by terraflops1048576** [comment]
After trying to fix the test failures that relate to the new exception info, I realized that this is actually rather problematic -- it fixes way too many errors that would otherwise be reported to the user, like the overflow error resulting from [1, 2**100], trying to concatenate `np.array([2**100])` and `np.array([1])`, etc. After all, it is perfectly legal to pickle all of these things and store it in a column, but we obviously don't want to do this for performance reasons. We would rather error and tell the user to adjust the problematic values.

I'm not sure what we should do with the np.array case, as they are technically objects, but maybe we shouldn't fall back if anything related to ArrowTensorArray fails.

File: python/ray/data/_internal/arrow_block.py
Comment (raulchen) [inline python/ray/data/_internal/arrow_block.py:70]: nit, it's not efficient to format the entire error as a string. 
Can we iterate over the causes and check their types?

File: python/ray/data/tests/test_transform_pyarrow.py
Comment (raulchen) [reply python/ray/data/tests/test_transform_pyarrow.py:188]: yes. it got run here https://buildkite.com/ray-project/microcheck/builds/2214

File: python/ray/data/_internal/arrow_ops/transform_pyarrow.py
Comment (raulchen) [reply python/ray/data/_internal/arrow_ops/transform_pyarrow.py:78]: Thanks for the detailed explanation! Can you also update the docstring of this method to reflect this change?
also do you mind explaining again what you mean by "set the promotion options differently"?

File: python/ray/air/util/object_extensions/arrow.py
Comment (raulchen) [reply python/ray/air/util/object_extensions/arrow.py:13]: This should be fine.  We'll consider bumping the minimum PyArrow version later.

**Review by raulchen — COMMENTED** [review]
Thanks for the new comments and explanation. 

> it fixes way too many errors that would otherwise be reported to the user, like the overflow error resulting from [1, 2**100], trying to concatenate np.array([2**100]) and np.array([1]), etc. After all, it is perfectly legal to pickle all of these things and store it in a column, but we obviously don't want to do this for performance reasons. 

To confirm my understanding, in the UDF return, if any value is overflowed, all values will fallback to objects. If nothing is overflowed, the type will still be integers. Right?

This seems fine. Because we don't know if the users intend to return overflowed integers either. Maybe we can introduce a flag to allow disabling the fallback.

> I'm not sure what we should do with the np.array case, as they are technically objects, but maybe we shouldn't fall back if anything related to ArrowTensorArray fails.

Are you concerned about np.arrays with object dtype? I think it's okay to not handle that. Users can just return a python list instead.

**Comment by raulchen** [comment]
Some tests are failing. can you take a look? https://buildkite.com/ray-project/microcheck/builds/2214

**Comment by terraflops1048576** [comment]
I've tried to separate out overflows and some other errors like mismatched float/int types and make them continue reporting errors as in the tests I've fixed, instead of saving them as objects in the PyArrow table, primarily just to avoid unintended consequences.

You are correct that if we enabled catching all ArrowConversionErrors and turning things into objects that as long as blocks were completely non-overflowed integers, they would be stored as integer type. Currently, np.ndarrays with object dtype (such as when a user tries to put in overflowed integers) are just pickled as objects, because it's really hard to detect whether it's the result of overflowed integers or just plain objects in there.

File: python/ray/data/_internal/arrow_block.py
Comment (raulchen) [inline python/ray/data/_internal/arrow_block.py:71]:
```suggestion

ARROW_OBJECT_FIXABLE_ERRORS = (
    pyarrow.lib.ArrowTypeError,
    pyarrow.lib.ArrowNotImplementedError,
    pyarrow.lib.ArrowInvalid,
)


def is_object_fixable_error(e: ArrowConversionError) -> bool:
    """Returns whether this error can be fixed by using an ArrowPythonObjectArray"""
    return isinstance(e.__cause__, ARROW_OBJECT_FIXABLE_ERRORS)
```

**Review by raulchen — APPROVED** [review]
Thanks for the updates. LGTM

**Comment by raulchen** [comment]
@terraflops1048576 the CI failures are unrelated and already fixed in master. Can you merge the latest master again?

**Comment by terraflops1048576** [comment]
I'm not sure why the tests that I fixed here weren't broken before, but I included a fix.

Interestingly enough, the [documentation](https://docs.ray.io/en/latest/data/api/doc/ray.data.Dataset.union.html) for `ray.data.Dataset.union` says

> The datasets must have the same schema as this dataset, otherwise the behavior is undefined.

However this isn't actually the case anymore and things like `python/ray/data/tests/test_consumption.py::test_union` actually depend on different behavior. Nevertheless, I fixed `unify_schemas` to reflect this behavior.

**Pull request merged**
